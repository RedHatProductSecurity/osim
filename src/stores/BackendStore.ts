import { ref } from 'vue';
import { defineStore } from 'pinia';


export interface OsimBackends {
  osidb: string | unknown;
  isReady: () => boolean;
  error: boolean | string;
}

export const useBackendStore = defineStore('BackendStore', () => {
  const error = ref<string>('');
  const osidb = ref<string>('');

  function isReady() {
    return osidb.value !== '';
  }

  return {
    error,
    osidb,
    isReady,
  };
});
