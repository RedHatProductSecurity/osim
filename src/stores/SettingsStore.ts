import { ref, watch } from 'vue';
import { defineStore } from 'pinia';
import { z } from 'zod';

import serviceWorkerClient from '../services/service-worker-client';


// Consideration: we may want some settings to be persistently stored,
//   but other settings must be in-memory only, e.g. api keys

const _settingsStoreKey = 'SettingsStore';


export const SettingsSchema = z.object({
  // bugzillaApiKey: z.string().length(
  //   32, {message: 'Bugzilla API Key is the wrong length!'}
  // ).optional(),
  bugzillaApiKey: z.string().optional().or(z.literal('')),
  jiraApiKey: z.string().optional().or(z.literal('')),
  showNotifications: z.boolean(),
});
export type SettingsType = z.infer<typeof SettingsSchema>;

// const settingsStoreSessionStorage = z.object({
//   settings: SettingsSchema,
// });
// type SettingsStoreSessionStorage = z.infer<typeof settingsStoreSessionStorage>;


export const useSettingsStore = defineStore('SettingsStore', () => {
  const settings = ref<SettingsType>({ showNotifications: false });
  // const settings = useSessionStorage(_settingsStoreKey, {} as SettingsType);
  serviceWorkerClient.listen(_settingsStoreKey, value => {
    const newSettingsStore = SettingsSchema.safeParse(value);
    if (newSettingsStore.success) {
      if (JSON.stringify(newSettingsStore.data) !== JSON.stringify(settings.value)) {
        // New value; update
        settings.value = newSettingsStore.data;
      }
    }
  });
  watch(settings, () => {
    serviceWorkerClient.put(_settingsStoreKey, JSON.parse(JSON.stringify(settings.value)));
  });

  function save(newSettings: SettingsType) {
    settings.value = { ...newSettings };
  }

  function $reset() {
    settings.value = { showNotifications: false };
  }

  return {
    $reset,
    save,
    settings,
  };
});

