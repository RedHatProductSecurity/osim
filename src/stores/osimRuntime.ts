// No need to use Pinia; these values are set on app startup and never changed.
// Named with camelCase instead of PascalCase to differentiate from Pinia stores.

import { computed, ref } from 'vue';
import { z } from 'zod';

// export enum OsimRuntimeDev {
//   PROD,
//   DEV,
//   MOCK,
// }
// export const osimRuntimeDev = OsimRuntimeDev.DEV;

export enum OsimRuntimeStatus {
  INIT,
  READY,
  ERROR,
}

const status = ref(OsimRuntimeStatus.INIT);
export const osimRuntimeStatus = computed<OsimRuntimeStatus>(() => {
  return status.value;
});

const OsimRuntime = z.object({
  backends: z.object({
    osidb: z.string(),
    bugzilla: z.string(),
  }),
  osimVersion: z.string(),
  error: z.string().default(''),
});
type OsimRuntime = z.infer<typeof OsimRuntime>;

const runtime = ref<OsimRuntime>({
  backends: { osidb: '', bugzilla: '' },
  osimVersion: '',
  error: '',
});
export const osimRuntime = computed<OsimRuntime>(() => {
  return runtime.value;
});


const OsidbHealthy = z.object({
  env: z.string(),
  revision: z.string(),
  version: z.string(),
});
type OsidbHealthy = z.infer<typeof OsidbHealthy>;
const _osidbHealth = ref<OsidbHealthy>({
  env: '',
  revision: '',
  version: '',
});
export const osidbHealth = computed<OsidbHealthy>(() => {
  return _osidbHealth.value;
});



let setupCallCount = 0;
export function setup() {
  if (setupCallCount > 0) {
    console.warn('osimRuntime.setup called more than once:', setupCallCount + 1);
  }
  setupCallCount++;

  return fetchRuntime()
    .then(fetchOsidbHealthy)
    .then(() => {
      // If an error was not set, then the status is still INIT.
      // If the status is still INIT after all the initialization, then set the status to READY.
      if (status.value === OsimRuntimeStatus.INIT) {
        status.value = OsimRuntimeStatus.READY;
      }
    });
}

function fetchRuntime() {
  return fetch('/runtime.json', {
    method: 'GET',
    cache: 'no-cache',
  })
    .then(response => response.json())
    .then((json: OsimRuntime) => {
      console.debug('OsimRuntime', json);
      try {
        runtime.value = OsimRuntime.parse(json);
      } catch (e) {
        console.error('Unable to parse OsimRuntime', e);
        runtime.value.error = 'Backends are not correctly configured. Please try again later.';
        status.value = OsimRuntimeStatus.ERROR;
      }
    })
    .catch(e => {
      console.error('Unable to get backends', e);
      runtime.value.error = 'Error finding backends. ' +
                            'Possible deployment misconfiguration. Please try again.';
      status.value = OsimRuntimeStatus.ERROR;
    });

}

function fetchOsidbHealthy() {
  return fetch(`${runtime.value.backends.osidb}/osidb/healthy`, {
    method: 'GET',
    cache: 'no-cache',
  })
    .then(response => response.json())
    .then((json: OsidbHealthy) => {
      console.debug('OsidbHealthy', json);
      try {
        _osidbHealth.value = OsidbHealthy.parse(json);
      } catch (e) {
        console.error('Unable to parse OSIDB health', e);
        runtime.value.error = 'Unexpected response from OSIDB. Please try again later.';
        status.value = OsimRuntimeStatus.ERROR;
      }
    })
    .catch(e => {
      console.error('Unable to get OSIDB health', e);
      runtime.value.error = 'Error getting OSIDB health. ' +
                            'Possible deployment misconfiguration. Please try again.';
      status.value = OsimRuntimeStatus.ERROR;
    });
}
