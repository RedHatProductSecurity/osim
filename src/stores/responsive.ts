// Track header and footer for elements between to have responsive positioning

import { ref } from 'vue';

const navbarHeight = ref<number>(0);
const navbarBottom = ref<number>(0);

const footerTop = ref<number>(0);
const footerHeight = ref<number>(0);

export {
  navbarHeight,
  navbarBottom,
  footerTop,
  footerHeight,
};
