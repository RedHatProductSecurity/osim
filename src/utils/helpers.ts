// Source: https://github.com/vuejs/core/issues/5303#issuecomment-1543596383

import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );
