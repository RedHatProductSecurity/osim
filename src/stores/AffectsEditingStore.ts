import { ref, computed, onUnmounted, type Ref } from 'vue';

import { clone } from 'ramda';
import { defineStore } from 'pinia';

import { useFlaw } from '@/composables/useFlaw';
import { useFlawAffectsModel } from '@/composables/useFlawAffectsModel';

import type { ZodAffectType } from '@/types/zodAffect';
import { matcherForAffect } from '@/utils/helpers';

export const useAffectsEditingStore = defineStore('EditableAffects', () => {
  const { flaw } = useFlaw();

  const { initialAffects } = useFlawAffectsModel();
  const affectSelections = ref<ZodAffectType[]>([]);

  const {
    areAllAffectsSelectable,
    areAllAffectsSelected,
    isAffectSelected,
    isIndeterminateSelection,
    isSelectable,
    resetSelections,
    selectedAffects,
    toggleAffectSelection,
    toggleMultipleAffectSelections,
  } = useAffectSelections(affectSelections, isBeingEdited);

  const affectValuesPriorEdit = ref<ZodAffectType[]>([]);
  const affectsBeingEdited = ref<ZodAffectType[]>([]);

  onUnmounted($reset);

  function isBeingEdited(affect: ZodAffectType) {
    const matchingAffect = affectsBeingEdited.value.find(matcherForAffect(affect)) as ZodAffectType;
    return affectsBeingEdited.value.includes(matchingAffect);
  }

  function getAffectPriorEdit(affect: ZodAffectType): ZodAffectType {
    return affectValuesPriorEdit.value.find(prior => prior.uuid && prior.uuid === affect.uuid) || affect;
  }

  function editAffect(affect: ZodAffectType) {
    if (isAffectSelected(affect)) {
      toggleAffectSelection(affect);
    }

    if (!isBeingEdited(affect)) {
      const affectToEdit = flaw.value.affects.find(matcherForAffect(affect));
      if (affectToEdit) {
        affectsBeingEdited.value.push(affectToEdit);
        const affectCopy = clone(affect) as ZodAffectType;
        affectValuesPriorEdit.value.push(affectCopy);
      }
    }
  }

  function commitChanges(affect: ZodAffectType) {
    const matchAffect = matcherForAffect(affect);
    const updateIndex = flaw.value.affects.findIndex(matchAffect);
    flaw.value.affects.splice(updateIndex, 1, affect);
    resetStagedAffectEdit(affect);
  }

  function cancelChanges(affect: ZodAffectType) {
    resetStagedAffectEdit(affect);
  }

  function revertAffectToLastSaved(affect: ZodAffectType) {
    const saved = initialAffects.value.find(a => a.uuid === affect.uuid);
    const index = flaw.value.affects.findIndex(a => a.uuid === affect.uuid);
    if (index !== -1 && saved) {
      flaw.value.affects[index] = { ...saved };
    }
    if (isAffectSelected(affect)) {
      toggleAffectSelection(affect);
    }
  }

  function resetStagedAffectEdit(affect: ZodAffectType) {
    const matchAffect = matcherForAffect(affect);
    const editingIndex = affectsBeingEdited.value.findIndex(matchAffect);
    affectsBeingEdited.value.splice(editingIndex, 1);
    affectValuesPriorEdit.value.splice(affectValuesPriorEdit.value.findIndex(matchAffect), 1);
    if (isAffectSelected(affect)) {
      toggleAffectSelection(affect);
    }
  }
  function editSelectedAffects() {
    selectedAffects.value.forEach(editAffect);
  }

  function commitAllChanges() {
    const affectsToCommit = [...affectsBeingEdited.value];
    affectsToCommit.forEach(commitChanges);
  }

  function cancelAllChanges() {
    const affectsToCancel = [...affectsBeingEdited.value];
    affectsToCancel.forEach(cancelChanges);
  }

  function $reset() {
    affectsBeingEdited.value = [];
    affectValuesPriorEdit.value = [];
  }

  return {
    cancelChanges,
    commitChanges,
    isBeingEdited,
    getAffectPriorEdit,
    editAffect,
    revertAffectToLastSaved,
    affectValuesPriorEdit,
    affectsBeingEdited,
    $reset,
    selectedAffects,
    areAllAffectsSelected,
    isIndeterminateSelection,
    isSelectable,
    areAllAffectsSelectable,
    isAffectSelected,
    toggleAffectSelection,
    toggleMultipleAffectSelections,
    resetSelections,
    editSelectedAffects,
    commitAllChanges,
    cancelAllChanges,
  };
});

function useAffectSelections(selectedAffects: Ref<ZodAffectType[]>, isBeingEdited: (affect: ZodAffectType) => boolean) {
  const { flaw } = useFlaw();
  const affects = computed(() => flaw.value.affects);

  const { isAffectBeingRemoved } = useFlawAffectsModel();

  function isSelectable(affect: ZodAffectType) {
    return !isAffectBeingRemoved(affect) && !isBeingEdited(affect);
  }

  function isAffectSelected(affect: ZodAffectType) {
    return isSelectable(affect) && selectedAffects.value.includes(affect);
  }

  const areAllAffectsSelected = computed(() => {
    return affects.value.every(isAffectSelected);
  });

  const isIndeterminateSelection = computed(() => {
    return !areAllAffectsSelected.value && affects.value.some(isAffectSelected);
  });

  const areAllAffectsSelectable = computed(() => affects.value.every(isSelectable));

  function toggleAffectSelection(affect: ZodAffectType) {
    if (!isSelectable(affect)) {
      return;
    }
    if (!isAffectSelected(affect)) {
      selectedAffects.value.push(affect);
    } else {
      selectedAffects.value = selectedAffects.value.filter(a => a !== affect);
    }
  }

  function toggleMultipleAffectSelections(event: Event) {
    if (areAllAffectsSelected.value) {
      affects.value.forEach(affect => toggleAffectSelection(affect));
    } else if (selectedAffects.value.length === 0) {
      affects.value.filter(affect => !isAffectSelected(affect)).forEach(affect => toggleAffectSelection(affect));
    } else {
      // De-select all selected affects if only some are selected
      affects.value.filter(affect => isAffectSelected(affect)).forEach(toggleAffectSelection);
      (event.target as HTMLInputElement).checked = false;
    }
  }

  function resetSelections() {
    selectedAffects.value = [];
  }

  return {
    selectedAffects,
    areAllAffectsSelected,
    isIndeterminateSelection,
    isSelectable,
    areAllAffectsSelectable,
    isAffectSelected,
    toggleAffectSelection,
    toggleMultipleAffectSelections,
    resetSelections,
  };
}
