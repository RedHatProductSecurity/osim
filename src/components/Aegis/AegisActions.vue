<script setup lang="ts">
import { osimRuntime } from '@/stores/osimRuntime';

const props = withDefaults(defineProps<{
  canShowFeedback: boolean;
  canSuggest: boolean;
  featureFlag?: null | string;
  hasAppliedSuggestion: boolean;
  hasMultipleSuggestions: boolean;
  isSuggesting: boolean;
  selectedIndex: number;
  suggestions: string[];
  tooltipText: string;
}>(), {
  featureFlag: 'aiCweSuggestions',
});

const emit = defineEmits<{
  feedback: ['negative' | 'positive'];
  revert: [];
  selectSuggestion: [index: number];
  suggest: [];
}>();

function selectSuggestion(index: number) {
  emit('selectSuggestion', index);
}
</script>

<template>
  <div v-if="props.featureFlag && osimRuntime.flags?.[props.featureFlag] === true" class="d-flex align-items-center">
    <i
      class="bi-stars label-icon"
      :class="{ disabled: !canSuggest, applied: hasAppliedSuggestion }"
      :title="tooltipText"
      @click.prevent.stop="canSuggest && emit('suggest')"
    />
    <div
      v-if="hasMultipleSuggestions && !isSuggesting"
      class="dropdown ms-1"
    >
      <i
        class="bi-chevron-down label-icon dropdown-toggle"
        data-bs-toggle="dropdown"
        data-bs-auto-close="outside"
        aria-expanded="false"
        title="Select from multiple suggestions"
      />
      <ul class="dropdown-menu suggestion-dropdown">
        <li>
          <h6 class="dropdown-header">AI Suggestions</h6>
        </li>
        <li>
          <hr class="dropdown-divider">
        </li>
        <slot
          name="suggestion-item"
          :suggestions="suggestions"
          :selected-index="selectedIndex"
          :select-suggestion="selectSuggestion"
        >
          <li v-for="(suggestion, index) in suggestions" :key="index">
            <button
              class="dropdown-item"
              :class="{ 'active': index === selectedIndex }"
              type="button"
              @click="selectSuggestion(index)"
            >
              <span class="fs-7">{{ suggestion }}</span>
              <span v-if="index === selectedIndex" class="text-success">
                <i class="bi-check-square-fill fs-5"></i>
              </span>
            </button>
          </li>
        </slot>
      </ul>
    </div>
    <span v-if="canShowFeedback && !isSuggesting" class="ms-2">
      <i
        class="bi-arrow-counterclockwise label-icon"
        title="Revert to previous value"
        @click.prevent.stop="emit('revert')"
      />
      <i
        class="bi-hand-thumbs-up label-icon"
        title="Mark suggestion helpful"
        @click.prevent.stop="emit('feedback', 'positive')"
      />
      <i
        class="bi-hand-thumbs-down label-icon"
        title="Mark suggestion unhelpful"
        @click.prevent.stop="emit('feedback', 'negative')"
      />
    </span>
  </div>
</template>

<style scoped lang="scss">
.label-icon {
  color: gray;
  margin-right: 0.5rem;
  cursor: pointer;

  &.disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }

  &.applied {
    color: black;
  }

  &:hover:not(.disabled) {
    color: #333;
  }
}

.suggestion-dropdown {
  min-width: 200px;
  max-height: 300px;
  overflow-y: auto;
  font-size: 0.875rem;

  .dropdown-header {
    padding: 0.25rem 0.5rem;
    font-size: 0.75rem;
  }

  :deep(.dropdown-item) {
    padding: 0.35rem 0.5rem;
    border-bottom: 1px solid #f8f9fa;
    display: flex;
    align-items: center;
    gap: 0.5rem;

    &:last-child {
      border-bottom: none;
    }

    &.active {
      background-color: #d1f0d4;
      border-left: 3px solid #198754;
      color: #212529;
    }

    &:active,
    &:focus {
      background-color: #d1f0d4 !important;
      color: #212529 !important;
    }
  }
}
</style>
