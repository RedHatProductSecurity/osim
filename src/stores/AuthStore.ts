import { computed, ref, watch } from 'vue';

import { defineStore } from 'pinia';
import { z } from 'zod';
import jwtDecode from 'jwt-decode';
import type { JwtPayload } from 'jwt-decode';
import { useLocalStorage } from '@vueuse/core';
import { DateTime } from 'luxon';

import router from '@/router';
import { osimRuntime } from '@/stores/osimRuntime';

function getDevRefreshToken(): null | string {
  return osimRuntime.value.env === 'dev' ? (localStorage.getItem('AuthStore::refresh') || null) : null;
}

function setDevRefreshToken(token: string) {
  if (osimRuntime.value.env === 'dev') {
    localStorage.setItem('AuthStore::refresh', token);
  }
}

const loginResponse = z.object({
  access: z.string(),
  refresh: z.string().optional(),
  env: z.string(),
  detail: z.string().optional(),
});

export const useAuthStore = defineStore('AuthStore', () => {
  const accessToken = ref<null | string>();
  const refreshToken = ref<null | string>();

  // Persist only login state; tokens remain in-memory except dev refresh token (handled separately)
  const isLoggedIn = useLocalStorage('AuthStore::isLoggedIn', false);

  function $reset() {
    accessToken.value = null;
    refreshToken.value = null;
    isLoggedIn.value = false;
    // Clear dev refresh token storage
    localStorage.removeItem('AuthStore::refresh');
  }

  function isAccessTokenExpired() {
    try {
      const exp = accessToken.value ? jwtDecode<JwtPayload>(accessToken.value)?.exp : null;
      return !exp || DateTime.now().toSeconds() >= exp - 60;
    } catch (e) {
      console.debug('AuthStore: access token not a valid JWT', accessToken.value, e);
      return true;
    }
  }

  async function login(username: string = '', password: string = '') {
    const requestMetadata: { [key: string]: any } = {};
    if (osimRuntime.value.backends.osidbAuth == 'kerberos') {
      requestMetadata['method'] = 'GET';
    } else if (osimRuntime.value.backends.osidbAuth == 'credentials') {
      requestMetadata['method'] = 'POST';
      requestMetadata['body'] = JSON.stringify({ username: username, password: password });
      requestMetadata['headers'] = {
        'Content-Type': 'application/json',
      };
    }

    return fetch(`${osimRuntime.value.backends.osidb}/auth/token`, {
      credentials: 'include',
      cache: 'no-cache',
      ...requestMetadata,
    })
      .then(async (response) => {
        if (!response.ok) {
          const text = await response.text();
          throw new Error(text);
        }
        const json = await response.json();
        const parsedLoginResponse = loginResponse.passthrough().parse(json);
        if (parsedLoginResponse.detail) {
          throw new Error(parsedLoginResponse.detail);
        }
        // Store refresh token policy: dev in localStorage, non-dev in memory
        if (parsedLoginResponse.refresh) {
          if (osimRuntime.value.env === 'dev') {
            setDevRefreshToken(parsedLoginResponse.refresh);
            refreshToken.value = null;
          } else {
            refreshToken.value = parsedLoginResponse.refresh;
          }
        } else {
          refreshToken.value = null;
        }
        accessToken.value = parsedLoginResponse.access;
        isLoggedIn.value = true;
        // expose env to UserStore for persistence/compatibility
        const { useUserStore } = await import('./UserStore');
        useUserStore().setEnv(parsedLoginResponse.env);
        return parsedLoginResponse.access;
      })
      .then((access) => {
        // Fetch whoami and update user store without creating a static import cycle
        return fetch(`${osimRuntime.value.backends.osidb}/osidb/whoami`, {
          credentials: 'include',
          cache: 'no-cache',
          headers: {
            Authorization: `Bearer ${access}`,
          },
        });
      })
      .then(response => response.json())
      .then(async (json) => {
        const { whoamiResponse } = await import('./UserStore');
        const parsedWhoamiResponse = whoamiResponse.parse(json);
        const { useUserStore } = await import('./UserStore');
        const userStore = useUserStore();
        userStore.setWhoami(parsedWhoamiResponse);
      })
      .then(async () => {
        const { useUserStore } = await import('./UserStore');
        const userStore = useUserStore();
        await userStore.updateJiraUsername();
      })
      .catch((e) => {
        $reset();
        console.error('AuthStore::login() unsuccessful login request', e);
        throw e;
      });
  }

  function logout() {
    $reset();
    return router.push({ name: 'login' });
  }

  const isAuthenticated = computed<boolean>(() => {
    if (!accessToken.value) {
      return false;
    }
    try {
      const decoded = jwtDecode<JwtPayload>(accessToken.value);
      const now = Date.now() / 1000;
      return decoded.exp != null && now < decoded.exp;
    } catch (e) {
      return false;
    }
  });

  // React to authentication changes similarly to previous UserStore behavior
  watch(isAuthenticated, async () => {
    if (isAuthenticated.value) {
      if (router.currentRoute.value.name === 'login') {
        try {
          const { queryRedirect } = await import('./UserStore');
          const maybeRedirect = queryRedirect.parse(router.currentRoute.value);
          const redirect = maybeRedirect.query.redirect;
          if (redirect.startsWith('/')) {
            router.push(redirect);
            return;
          } else {
            console.debug('AuthStore::isAuthenticated() Refusing to redirect to', redirect);
          }
        } catch (e) {
          // do nothing
        }
        router.push({ name: 'index' });
        return;
      }
    } else {
      // Clear on unauthenticated and redirect to login
      $reset();
      if (router.currentRoute.value.name !== 'login') {
        const currentPath = router.currentRoute.value.fullPath;
        if (currentPath !== '/') {
          const query: any = {};
          query.redirect = currentPath;
          router.push({ name: 'login', query });
          return;
        }
        router.push({ name: 'login' });
        return;
      }
    }
  });

  return {
    isLoggedIn,
    accessToken,
    refreshToken,
    isAccessTokenExpired,
    isAuthenticated,
    getDevRefreshToken,
    setDevRefreshToken,
    login,
    logout,
    $reset,
  };
});
