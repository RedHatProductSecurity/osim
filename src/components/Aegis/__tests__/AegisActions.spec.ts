import { describe, it, expect, beforeEach } from 'vitest';

import AegisActions from '@/components/Aegis/AegisActions.vue';

import { mountWithConfig } from '@/__tests__/helpers';
import { osimRuntime } from '@/stores/osimRuntime';

describe('aegisActions', () => {
  beforeEach(() => {
    // @ts-expect-error - osimRuntime is readonly in tests
    osimRuntime.value = {
      ...osimRuntime.value,
      flags: {
        ...osimRuntime.value.flags,
        aiCweSuggestions: true,
      },
    };
  });

  it('renders correctly', () => {
    const wrapper = mountWithConfig(AegisActions, {
      props: {
        canShowFeedback: false,
        canSuggest: true,
        hasAppliedSuggestion: false,
        hasMultipleSuggestions: false,
        isSuggesting: false,
        selectedIndex: 0,
        suggestions: [],
        tooltipText: 'Test tooltip',
      },
    });

    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('emits suggest event when star icon clicked', async () => {
    const wrapper = mountWithConfig(AegisActions, {
      props: {
        canShowFeedback: false,
        canSuggest: true,
        hasAppliedSuggestion: false,
        hasMultipleSuggestions: false,
        isSuggesting: false,
        selectedIndex: 0,
        suggestions: [],
        tooltipText: 'Suggest CWE',
      },
    });

    await wrapper.find('.bi-stars').trigger('click');
    expect(wrapper.emitted('suggest')).toHaveLength(1);
  });

  it('shows dropdown when multiple suggestions available', () => {
    const wrapper = mountWithConfig(AegisActions, {
      props: {
        canShowFeedback: false,
        canSuggest: true,
        hasAppliedSuggestion: true,
        hasMultipleSuggestions: true,
        isSuggesting: false,
        selectedIndex: 0,
        suggestions: ['CWE-79', 'CWE-89'],
        tooltipText: 'AI Suggestions',
      },
    });

    expect(wrapper.find('.bi-chevron-down').exists()).toBe(true);
  });

  it('does not show dropdown when only one suggestion', () => {
    const wrapper = mountWithConfig(AegisActions, {
      props: {
        canShowFeedback: false,
        canSuggest: true,
        hasAppliedSuggestion: true,
        hasMultipleSuggestions: false,
        isSuggesting: false,
        selectedIndex: 0,
        suggestions: ['CWE-79'],
        tooltipText: 'AI Suggestions',
      },
    });

    expect(wrapper.find('.bi-chevron-down').exists()).toBe(false);
  });

  it('shows feedback actions when suggestion applied', () => {
    const wrapper = mountWithConfig(AegisActions, {
      props: {
        canShowFeedback: true,
        canSuggest: false,
        hasAppliedSuggestion: true,
        hasMultipleSuggestions: false,
        isSuggesting: false,
        selectedIndex: 0,
        suggestions: ['CWE-79'],
        tooltipText: 'AI Suggestions',
      },
    });

    expect(wrapper.find('.bi-arrow-counterclockwise').exists()).toBe(true);
    expect(wrapper.find('.bi-hand-thumbs-up').exists()).toBe(true);
    expect(wrapper.find('.bi-hand-thumbs-down').exists()).toBe(true);
  });

  it('emits feedback events', async () => {
    const wrapper = mountWithConfig(AegisActions, {
      props: {
        canShowFeedback: true,
        canSuggest: false,
        hasAppliedSuggestion: true,
        hasMultipleSuggestions: false,
        isSuggesting: false,
        selectedIndex: 0,
        suggestions: ['CWE-79'],
        tooltipText: 'AI Suggestions',
      },
    });

    await wrapper.find('.bi-arrow-counterclockwise').trigger('click');
    expect(wrapper.emitted('revert')).toHaveLength(1);

    await wrapper.find('.bi-hand-thumbs-up').trigger('click');
    expect(wrapper.emitted('feedback')).toHaveLength(1);
    expect(wrapper.emitted('feedback')?.[0]).toEqual(['positive']);

    await wrapper.find('.bi-hand-thumbs-down').trigger('click');
    expect(wrapper.emitted('feedback')).toHaveLength(2);
    expect(wrapper.emitted('feedback')?.[1]).toEqual(['negative']);
  });

  it('disables suggest icon when canSuggest is false', () => {
    const wrapper = mountWithConfig(AegisActions, {
      props: {
        canShowFeedback: false,
        canSuggest: false,
        hasAppliedSuggestion: false,
        hasMultipleSuggestions: false,
        isSuggesting: false,
        selectedIndex: 0,
        suggestions: [],
        tooltipText: 'Cannot suggest',
      },
    });

    const starIcon = wrapper.find('.bi-stars');
    expect(starIcon.classes()).toContain('disabled');
  });
});
