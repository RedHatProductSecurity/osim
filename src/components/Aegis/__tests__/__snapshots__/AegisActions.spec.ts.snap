// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`aegisActions > renders correctly 1`] = `
"<div data-v-e049df0b="" class="d-flex align-items-center"><i data-v-e049df0b="" class="bi-stars label-icon" title="Test tooltip"></i>
  <!--v-if-->
  <!--v-if-->
</div>"
`;
